nixos/modules/tasks/swraid.nix

```diff
@@ -1,17 +1,50 @@
-{ pkgs, ... }:
+{ pkgs, lib, config, ... }:
 
+with lib;
+let cfg = config.services.mdadm;
+in
 {
 
-  environment.systemPackages = [ pkgs.mdadm ];
+  options = {
 
-  services.udev.packages = [ pkgs.mdadm ];
+    services.mdadm = {
 
-  systemd.packages = [ pkgs.mdadm ];
+      monitor = {
+        enable = mkEnableOption "Enable the mdadm monitoring daemon";
+      };
 
-  boot.initrd.availableKernelModules = [ "md_mod" "raid0" "raid1" "raid10" "raid456" ];
+      adminAddr = mkOption {
+        type = types.str;
+        example = "admin@example.org";
+        description = "E-mail address of the server administrator.";
+      };
+    };
+  };
 
-  boot.initrd.extraUdevRulesCommands = ''
-    cp -v ${pkgs.mdadm}/lib/udev/rules.d/*.rules $out/
-  '';
+  config = mkMerge [
+    {
+      environment.systemPackages = [ pkgs.mdadm ];
+
+      services.udev.packages = [ pkgs.mdadm ];
+
+      systemd.packages = [ pkgs.mdadm ];
+
+      boot.initrd.availableKernelModules = [ "md_mod" "raid0" "raid1" "raid10" "raid456" ];
+
+      boot.initrd.extraUdevRulesCommands = ''
+        cp -v ${pkgs.mdadm}/lib/udev/rules.d/*.rules $out/
+      '';
+    }
+
+    (mkIf cfg.monitor.enable
+        {
+          systemd.services.mdadm-monitor = {
+            description = "Monitor RAID disks";
+            wantedBy = [ "multi-user.target" ];
+            script = "${pkgs.mdadm}/bin/mdadm --monitor -m ${assert cfg.adminAddr != null; cfg.adminAddr} --scan";
```

Comment:
This should use a NixOS assertion rather than a raw `assert`, as it makes the user experience better. You can find more information about NixOS assertions here: https://nixos.org/manual/nixos/stable/index.html#sec-assertions